Bug fix, has_many restrict_with_error translation should be set on parent

alex-handley opened the pull request:
When you want to set a translation on a has_many relation with its dependent set to restrict_with_error you have to set it on the child instead of the parent e.g

```
en.activerecord.attributes.company.companies
```

When it should be 

```
en.activerecord.attributes.restricted_with_error_firm.companies
```

This was also reported in #17003

egilburg commented on activerecord/test/cases/associations/has_many_associations_test.rb:
```diff
@@ -1441,6 +1441,31 @@ def test_restrict_with_error
   end
 
+  def test_restrict_with_error_and_translation
+    translation = { :activerecord => {
```
please use 1.9 hash syntax ( `key: value` over `:key => value`)

alex-handley replied:
@egilburg Sure, I used 1.8 syntax because the rest of the file is styled like this

alex-handley commented:
Before I push the hash syntax change does the Rails team have a preference on hash formatting.
Should it follow the format of the file or just use the new syntax?

kaspth commented:
@alex-handley Always new syntax :smile:

matthewd commented:
(but only on the newly added code; no need to reformat the rest of the file)

alex-handley commented:
Thanks! I have updated the syntax . Are you happy with the change?

arthurnn commented on activerecord/test/cases/associations/has_many_associations_test.rb:
```diff
@@ -1441,6 +1441,31 @@ def test_restrict_with_error
+                      }
+                    }
+                  }
+    I18n.backend.store_translations("en", translation)
```
the cleanup should happen in a `ensure` block I would say.

alex-handley replied:
Thanks! I wasn't sure how to do that in testunit. I have updated.

alex-handley committed:
has_many restrict_with_error translation should be set on parent

Currently when you want to set a translation on a has_many relation you
have to set the translation on the child model e.g. (from the test models)

en:
  activerecord:
    attributes:
      companies:
        companies: "public company name"

Instead of:

en:
  activerecord:
    attributes:
      restricted_with_error_firm:
        companies: "public company name"

maclover7 commented:
Please rebase. Also, it looks like the tests are failing on Travis CI.

alex-handley commented:
@maclover7 I have rebased and the spec are now back to green :)

vipulnsward commented on activerecord/test/cases/associations/has_many_associations_test.rb:
```diff
@@ -1516,6 +1516,32 @@ def test_restrict_with_error_with_locale
+    I18n.backend.store_translations("en", translation)
+
+    firm = RestrictedWithErrorFirm.create!(name: "restrict")
+    firm.companies.create(name: "child")
```
`create!`

vipulnsward commented:
This should have a changelog.

alex-handley commented:
@vipulnsward Rebased and added a changelog.

arthurnn commented:
Not sure if the regression test is not correct, or the fix.
Because if I revert the fix only, the test still passes. So one of them is not correct.

alex-handley commented:
Closing as this is now over a year old and stale.
